frontend/streamlit_app.py

```diff
@@ -0,0 +1,159 @@
+
+    Returns:
+    - tuple: Contains the AI response and sources.
+    """
+    url = os.getenv("CHAT_ENDPOINT", "http://localhost:8000/chatApp")
+
+    headers = {"accept": "application/json", "Content-Type": "application/json"}
+
+    payload = {"query": user_input, "listSources": True, "listContext": True}
+
+    try:
+        response = requests.post(url, headers=headers, json=payload)
+        response.raise_for_status()
+
+        try:
+            data = response.json()
+            if not isinstance(data, dict):
+                st.error("Invalid response format")
+                return None, None
+
+        except ValueError:
+            st.error("Failed to decode JSON response")
+            return None, None
+
+        sources = data.get("sources", "")
+        st.session_state.metadata[user_input] = {
+            "sources": sources,
+            "context": data.get("context", ""),
+        }
+
+        return data.get("response", ""), sources
+
+    except requests.exceptions.RequestException as e:
+        st.error(f"Request failed: {e}")
+        return None, None
+
+def main():
+    load_dotenv()
+
+    img = Image.open("assets/or_logo.png")
+    st.set_page_config(page_title="OR Assistant", page_icon=img)
+
+    deployment_time = datetime.datetime.now(pytz.timezone("UTC"))
+    st.info(f'Deployment time: {deployment_time.strftime("%m/%d/%Y %H:%M")} UTC')
+
+    st.title("OR Assistant")
+
+    if not os.getenv("CHAT_ENDPOINT"):
+        st.warning(
+            "The CHAT_ENDPOINT environment variable is not set or is empty. DEFAULT: http://localhost:8000/chatApp"
+        )
+
+    if "chat_history" not in st.session_state:
+        st.session_state.chat_history = []
+    if "metadata" not in st.session_state:
+        st.session_state.metadata = {}
+
+    if not st.session_state.chat_history:
+        st.session_state.chat_history.append({
+            "content": "Hi, I am the OpenROAD assistant. Type your query about OpenROAD",
+            "role": "ai",
+        })
+
+    for message in st.session_state.chat_history:
+        with st.chat_message(message["role"]):
+            st.markdown(message["content"])
+
+    user_input = st.chat_input("Enter your queries ...")
+
+    if user_input:
+        st.session_state.chat_history.append({"content": user_input, "role": "user"})
+
+        with st.chat_message("user"):
+            st.markdown(user_input)
+
+        response, sources = response_generator(user_input)
+        if response is not None:
+            response_buffer = ""
+
+            with st.chat_message("ai"):
+                message_placeholder = st.empty()
+
+                response_buffer = ""
+                for chunk in response.split(' '):
+                    response_buffer += chunk + ' '
+                    if chunk.endswith('\n'):
+                        response_buffer += ' '
+                    message_placeholder.markdown(response_buffer)
+                    time.sleep(0.05)
```

Comment:
Why is this needed?

Reply:
Are you trying to simulate streaming mode? This might help: https://github.com/langchain-ai/chat-langchain/issues/39

In fact, for long responses (100 words) you incur up to 5s delay. The right way is to enable streaming on the LLM endpoint, then process the output. What you have here is you generate the whole output, and then wait 0.05s per word. Which adds delay

Reply:
This isn't needed; it's just a visual thing. It's inefficient, as you said. It was initially made with @palaniappan-r 's endpoint structure in mind. Now, all the features such as human evaluation, auto evaluation, feedback, and frontend are dependent on the current API structure. For streaming to happen directly from LangChain, I will need to coordinate with Pal and create a separate /stream endpoint and test the possibility of it.